light-client/src/peer_list.rs

```diff
@@ -161,13 +211,110 @@ impl PeerListBuilder {
     /// - A primary has been set with a call to `PeerListBuilder::primary`.
     #[pre(self.primary.is_some())]
     #[post(PeerList::invariant(&ret))]
-    pub fn build(self) -> PeerList {
+    pub fn build(self) -> PeerList<T> {
         PeerList {
-            instances: self.instances,
+            values: self.values,
             primary: self.primary.unwrap(),
             witnesses: self.witnesses,
             full_nodes: self.full_nodes,
             faulty_nodes: self.faulty_nodes,
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    trait BTreeSetExt<T> {
+        fn to_vec(&self) -> Vec<T>;
+    }
+
+    impl BTreeSetExt<PeerId> for BTreeSet<PeerId> {
+        fn to_vec(&self) -> Vec<PeerId> {
+            self.iter().copied().collect()
+        }
+    }
+
+    fn a() -> PeerId {
```

Comment:
You could name these according to their use case further down.

Reply:
@xla Thanks for the review! :) See my answers inline.

> This is a great refactoring and I agree it makes the core of the verification indeed easier to follow and comprehend. My main concern is the very constrained T on the PeerList. Could it not give stronger guarantees if Instance would be a trait and PeerList expects that as its type parameter?

I don't really see the value of having `Instance` defined as a trait and constraining `T` over that, since the behavior of `PeerList` does not depend in any way on the choice of `T`. My general approach to building data structures is to keep them as generic as possible, and only constrain on types if absolutely required by the definition of the struct, or on a per-method basis. Admittedly, this is a mindset I inherited from working in Haskell, but sees shared by the Rust community as well: https://github.com/rust-lang/rust-clippy/issues/1689.

Of course, if there is actual (type) safety to be gained by having a bound on `T` (that I am at the moment not seeing, but feel free to expand on that), I'd be happy to revisit that decision in that case.